File: Ejercicios Propuestos 3/Ejercicio 16/Ejercicio 16.py
NAME = "Renzo Leonardo Cienfuegos Cardenas"
GROUP = 2

# Función para realizar la partición del vector
def particion(vector, inicio, fin):

    pivote = vector[fin]
    i = inicio - 1

    for j in range(inicio, fin):

        if vector[j] <= pivote:
            i += 1

            # Intercambiar elementos
            vector[i], vector[j] = vector[j], vector[i]

    # Colocar el pivote en su posición correcta
    vector[i + 1], vector[fin] = vector[fin], vector[i + 1]

    return i + 1


# Función recursiva QuickSort
def quick_sort(vector, inicio, fin):

    if inicio < fin:

        pivote = particion(vector, inicio, fin)

        quick_sort(vector, inicio, pivote - 1)
        quick_sort(vector, pivote + 1, fin)


# Programa principal

n = int(input("Ingrese la cantidad de elementos: "))

vector = []

for i in range(n):
    numero = int(input(f"Elemento {i}: "))
    vector.append(numero)

# Ordenar el vector
quick_sort(vector, 0, n - 1)

# Mostrar el resultado
print("\nVector ordenado:")

for numero in vector:
    print(numero, end=" ")

print()

File: Ejercicios Propuestos 3/Ejercicio 16/ejercicio16.cpp
#include <iostream>
using namespace std;

int particion(int v[], int inicio, int fin) {
    int pivote = v[fin];
    int i = inicio - 1;

    for (int j = inicio; j < fin; j++) {
        if (v[j] <= pivote) {
            i++;
            int temp = v[i];
            v[i] = v[j];
            v[j] = temp;
        }
    }
    int temp = v[i + 1];
    v[i + 1] = v[fin];
    v[fin] = temp;

    return i + 1;
}

void quickSort(int v[], int inicio, int fin) {
    if (inicio < fin) {
        int p = particion(v, inicio, fin);
        quickSort(v, inicio, p - 1);
        quickSort(v, p + 1, fin);
    }
}

int main() {
    int n;
    cout << "Ingrese la cantidad de elementos: ";
    cin >> n;

    int v[100];
    for (int i = 0; i < n; i++) {
        cout << "Elemento " << i << ": ";
        cin >> v[i];
    }

    quickSort(v, 0, n - 1);

    cout << "Vector ordenado: ";
    for (int i = 0; i < n; i++) {
        cout << v[i] << " ";
    }
    cout << endl;

    return 0;
}
